gst/nnstreamer/tensor_query/tensor_query_common.c

```diff
@@ -546,17 +514,66 @@ query_tcp_send (GSocket * socket, uint8_t * data, size_t size,
 {
   gsize bytes_sent = 0;
   gssize rret;
-  GError *err;
+  GError *err = NULL;
   while (bytes_sent < size) {
     rret = g_socket_send (socket, (gchar *) data + bytes_sent,
         size - bytes_sent, cancellable, &err);
+    if (rret == 0) {
+      nns_logi ("Connection closed");
+      return FALSE;
+    }
     if (rret < 0) {
       nns_loge ("Error while sending data %s", err->message);
-      g_clear_error (&err);
       return FALSE;
     }
     bytes_sent += rret;
   }
-  nns_logd ("sent %zu", bytes_sent);
   return TRUE;
 }
+
+/**
+ * @brief [TCP] Callback for socket listener that pushes socket to the queue
+ */
+static void
+accept_socket_async_cb (GObject * source, GAsyncResult * result,
+    gpointer user_data)
+{
+  GSocketListener *socket_listener = G_SOCKET_LISTENER (source);
+  GSocket *socket;
+  GSocketAddress *saddr;
+  GError *err = NULL;
+  TensorQueryServerData *sdata = user_data;
+  TensorQueryConnection *conn;
+
+  socket =
+      g_socket_listener_accept_socket_finish (socket_listener, result, NULL,
+      &err);
+  if (!socket) {
+    nns_loge ("Failed to get socket: %s", err->message);
+    return;
+  }
+
+  /* create socket with connection */
+  conn = g_try_new (TensorQueryConnection, 1);
+  if (!conn) {
+    nns_loge ("Failed to allocate connection");
+    return;
+  }
+  saddr = g_socket_get_remote_address (socket, &err);
+  if (!saddr) {
+    nns_loge ("Failed to get socket address: %s", err->message);
+    return;
+  }
+  conn->protocol = (g_socket_get_protocol (socket) == G_SOCKET_PROTOCOL_TCP) ?
+      _TENSOR_QUERY_PROTOCOL_TCP : _TENSOR_QUERY_PROTOCOL_END;
+  conn->host = g_inet_address_to_string (g_inet_socket_address_get_address (
+          (GInetSocketAddress *) saddr));
+  conn->port = g_inet_socket_address_get_port ((GInetSocketAddress *) saddr);
+  conn->socket = socket;
+  conn->cancellable = g_cancellable_new ();
+  nns_logd ("connected from %s:%u", conn->host, conn->port);
+  g_async_queue_push (sdata->conn_queue, conn);
+
+  g_socket_listener_accept_socket_async (socket_listener,
+      sdata->cancellable, (GAsyncReadyCallback) accept_socket_async_cb, sdata);
```

Comment:
Does callback register as a callback?

Reply:
Calling `g_socket_listener_accept_socket_finish` detaches the callback so I need to re-register itself